segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,377 @@
+            log.warn("Write called on a readonly handle of segment {}", handle.getSegmentName());
+            throw new IllegalArgumentException("Write called on a readonly handle of segment "
+                    + handle.getSegmentName());
+        }
+
+        //Fix for Jenkins as Jenkins runs as super user privileges.
+        if ( !isWritableFile(path)) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        long fileSize = path.toFile().length();
+        if (fileSize < offset) {
+            throw new BadOffsetException(handle.getSegmentName(), fileSize, offset);
+        } else {
+            try (FileChannel channel = FileChannel.open(path,
+                    StandardOpenOption.WRITE); ReadableByteChannel sourceChannel = Channels.newChannel(data)) {
+                while (length != 0) {
+                    long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                    offset += bytesWritten;
+                    length -= bytesWritten;
+                }
+            }
+            return null;
+        }
+    }
+
+    private boolean isWritableFile(Path path) throws IOException {
+        PosixFileAttributes attrs = null;
+            attrs = Files.readAttributes(path,
+                    PosixFileAttributes.class);
+            return attrs.permissions().contains(OWNER_WRITE);
+    }
+
+    @SneakyThrows
+    private Void syncSeal(SegmentHandle handle) {
+
+        if (handle.isReadOnly()) {
+            log.info("Seal called on a read handle for segment {}", handle.getSegmentName());
+            throw new IllegalArgumentException(handle.getSegmentName());
+        }
+            Set<PosixFilePermission> perms = new HashSet<>();
+            // add permission as r--r--r-- 444
+            perms.add(PosixFilePermission.OWNER_READ);
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            Files.setPosixFilePermissions(Paths.get(config.getFilesystemRoot(), handle.getSegmentName()), perms);
+            log.info("Successfully sealed segment {}", handle.getSegmentName());
+            return null;
+    }
+
+    @SneakyThrows(IOException.class)
+    private Void syncConcat(SegmentHandle targetHandle, long offset, String sourceSegment) {
+
+        Path sourcePath = Paths.get(config.getFilesystemRoot(), sourceSegment);
+        Path targetPath = Paths.get(config.getFilesystemRoot(), targetHandle.getSegmentName());
+
+        try (FileChannel targetChannel = (FileChannel) Files.newByteChannel(targetPath, EnumSet.of(StandardOpenOption.APPEND));
+             RandomAccessFile sourceFile = new RandomAccessFile(String.valueOf(sourcePath), "r")) {
+            if (isWritableFile(sourcePath)) {
+                throw new IllegalStateException(sourceSegment);
+            }
+
+            targetChannel.transferFrom(sourceFile.getChannel(), offset, sourceFile.length());
```

Comment:
You need to check the result of this, then run it in a loop.

**Have you performance tested this? This looks very suspicious and there is no guarantee that it's a native operation. If this goes through the client (and implicitly via the NFS socket), then this will kill performance.**